src/masternodes/validation.cpp

```diff
@@ -2486,6 +2488,45 @@ static Res ProcessEVMQueue(const CBlock &block, const CBlockIndex *pindex, CCust
     return Res::Ok();
 }
 
+static Res ProcessDST20Migration(const CBlockIndex *pindex, CCustomCSView &cache, const CChainParams& chainparams, const uint64_t evmQueueId) {
+    if (!IsEVMEnabled(pindex->nHeight, cache, chainparams.GetConsensus())) return Res::Ok();
+
+    CrossBoundaryResult result;
+    auto evmTargetBlock = evm_unsafe_try_get_target_block_in_q(result, evmQueueId);
+    if (!result.ok) return DeFiErrors::DST20MigrationFailure(result.reason.c_str());
+    if (evmTargetBlock > 0) return Res::Ok();
+
+    auto time = GetTimeMillis();
+    LogPrintf("DST20 migration ...\n");
+
+    cache.ForEachToken([&](DCT_ID const &id, CTokensView::CTokenImpl token) {
+        if (token.IsPoolShare()) return true;
+        if (!token.IsDAT()) return true;
+
+        if (evm_try_is_dst20_deployed_or_queued(result, evmQueueId, token.name, token.symbol, id.ToString())) {
+            return result.ok;
+        }
+
+        evm_try_create_dst20(result, evmQueueId, token.creationTx.GetByteArray(),
+                token.name, token.symbol, id.ToString());
+        return result.ok;
+    }, DCT_ID{1});  // start from non-DFI
+    if (!result.ok) return DeFiErrors::DST20MigrationFailure(result.reason.c_str());
+
+    cache.ForEachLoanToken([&](DCT_ID const & id, CLoanView::CLoanSetLoanTokenImpl loanToken) {
```

Comment:
loanTokens are DATs, right? so this second loop has no effect as all DATs are already deployed or queued from above.

Reply:
It should be but there's no check in attributes to assert that every loan tokens are DAT. 
This is just to err on the safe side but it shouldn't affect the overall migration since we guard against `is_dst20_deployed_or_queued`.

Reply:
Fixed in https://github.com/DeFiCh/ain/pull/2258/commits/fda389e0bc72650ba6c84246250ecdd4dd5faefe and https://github.com/DeFiCh/ain/pull/2258/commits/e30f1faf15ea901c47ba28b69660f198695a1ea4